src/openvic-simulation/utility/Utility.hpp

```diff
@@ -11,4 +15,48 @@ namespace OpenVic::utility {
 		__assume(false);
 #endif
 	}
+
+	template<class T>
+	constexpr inline void hash_combine(std::size_t& s, const T& v) {
+		std::hash<T> h;
+		s ^= h(v) + 0x9e3779b9 + (s << 6) + (s >> 2);
+	}
+
+	template<size_t Shift, class T>
+	constexpr inline void hash_combine_index(std::size_t& s, const T& v) {
+		std::hash<T> h;
+		if constexpr(Shift == 0) {
+			s = h(v);
+		} else {
+			s ^= h(v) << Shift;
+		}
+	}
+
+	template<class T, typename ...Args>
+	constexpr void perfect_hash(std::size_t& s, T&& v, Args&&... args) {
+		static_assert(sizeof(T) + (sizeof(Args) + ...) <= sizeof(std::size_t), "Perfect hashes must be able to fit into size_t");
+		std::hash<T> h;
+		if constexpr(sizeof...(args) == 0) {
+			s = h(v);
+		} else {
+			const std::tuple arg_tuple { args... };
+			s = h(v) << (sizeof(T) * CHAR_BIT);
```

Comment:
This simplifies the varadic template, doing it in other ways requires using the previous argument which is a pain with varadic templates, we have to shift over to make room for the argument about to be placed, then we OR it so that its bits are placed in the hash

Like say I have char 254, char 254, and short 32766, this is the "perfect_hash" formula:
`(((0b011111110 << 8) | 0b01111110) << 16) | 0b0111111111111110`
and the result is this:
`0b011111110011111100111111111111110`
which completely fills the bit space for every value.